packages/lodestar/src/chain/opPools/aggregatedAttestationPool.ts

```diff
@@ -0,0 +1,310 @@
+  private lowestPermissibleSlot = 0;
+
+  add(attestation: phase0.Attestation, attestingIndices: ValidatorIndex[], committee: ValidatorIndex[]): InsertOutcome {
+    const slot = attestation.data.slot;
+    const lowestPermissibleSlot = this.lowestPermissibleSlot;
+
+    // Reject any attestations that are too old.
+    if (slot < lowestPermissibleSlot) {
+      throw new OpPoolError({code: OpPoolErrorCode.SLOT_TOO_LOW, slot, lowestPermissibleSlot});
+    }
+    const attestationGroupByDataHash = this.attestationGroupByDataHashBySlot.getOrDefault(slot);
+    const dataRoot = ssz.phase0.AttestationData.hashTreeRoot(attestation.data);
+    const dataRootHex = toHexString(dataRoot);
+    let attestationGroup = attestationGroupByDataHash.get(dataRootHex);
+    if (attestationGroup) {
+      return attestationGroup.add({attestation, attestingIndices});
+    } else {
+      attestationGroup = new MatchingDataAttestationGroup(committee);
+      attestationGroup.add({attestation, attestingIndices});
+      attestationGroupByDataHash.set(dataRootHex, attestationGroup);
+      return InsertOutcome.NewData;
+    }
+  }
+
+  /** Remove attestations which are too old to be included in a block. */
+  prune(clockSlot: Slot): void {
+    // Only retain SLOTS_PER_EPOCH slots
+    pruneBySlot(this.attestationGroupByDataHashBySlot, clockSlot, SLOTS_PER_EPOCH);
+    this.lowestPermissibleSlot = Math.max(clockSlot - SLOTS_PER_EPOCH, 0);
+  }
+
+  /**
+   * Get attestations to be included in a block.
+   */
+  getAttestationsForBlock(
+    state: CachedBeaconState<allForks.BeaconState>,
+    isValidAttestation = false
+  ): phase0.Attestation[] {
+    const forkName = state.config.getForkName(state.slot);
+    if (forkName === ForkName.phase0) {
+      return this.getAttestationsForPhase0Block(state, isValidAttestation);
+    }
+    return this.getAttestationsForAltairBlock(state, isValidAttestation);
+  }
+
+  /**
+   * Get all attestations optionally filtered by `attestation.data.slot`
+   * @param bySlot slot to filter, `bySlot === attestation.data.slot`
+   */
+  getAll(state: CachedBeaconState<allForks.BeaconState>, bySlot?: Slot): phase0.Attestation[] {
+    // we don't want to validate attestations
+    let attestations = this.getAttestationsForBlock(state, true);
+    if (bySlot) {
+      attestations = attestations.filter((att) => att.data.slot === bySlot);
+    }
+    return attestations;
+  }
+
+  /**
+   * Get attestations to be included in a phase0 block.
+   * As we are close to altair, this is not really important, it's mainly for e2e.
+   * The performance is not great due to the different BeaconState data structure to altair.
+   */
+  private getAttestationsForPhase0Block(
+    state: CachedBeaconState<allForks.BeaconState>,
+    isValidAttestation = false
+  ): phase0.Attestation[] {
+    if (state.config.getForkName(state.slot) !== ForkName.phase0) {
+      throw new Error(`Not support proposing block at slot ${state.slot}`);
```

Comment:
Is it necessary to do this check again here after https://github.com/ChainSafe/lodestar/pull/2851/files#diff-d9b2c635e7fcd4e4b3d0bddf21deb28e7d1b5f16a0ee899de077347bcbba77adR66 ?